glogger/handler.py

```diff
@@ -0,0 +1,277 @@
+        # would be easier to miss and less flexible.
+        send = backoff.on_exception(backoff.expo, exception=RequestException, max_tries=self.max_send_tries)(
+            self._send_logs
+        )
+
+        self.last_flush_time = self.time_fn()
+        try:
+            batch, response = send()
+            response.raise_for_status()
+        except Exception:
+            logger.exception(SERVER_SEND_ERROR_MESSAGE)
+        else:
+            self._drop_sent_batch(batch)
+
+    def _send_logs(self) -> Tuple[Batch, requests.Response]:
+        """
+        :return: (batch that was sent, response)
+        """
+        # Upon every retry we will remake the batch, in case we are able to batch more messages together.
+        batch = self.make_batch()
+        batch_data = {
+            "batch_id": batch.ident,
+            "metadata": self.get_metadata(),
+            "lost": batch.lost,
+            "logs": "<LOGS_JSON>",
+        }
+        # batch.logs is a list of json strings so ",".join() it into the final json string instead of json-ing the list.
+        data = self.jsonify(batch_data).replace('"<LOGS_JSON>"', f"[{','.join(batch.logs)}]").encode("utf-8")
+        # Default compression level (9) is slowest. Level 6 trades a bit of compression for speed.
+        data = gzip.compress(data, compresslevel=6)
+        response = self.session.post(
+            self.uri,
+            data=data,
+            headers={
+                "Content-Encoding": "gzip",
+                "Content-Type": "application/json",
+            },
+            timeout=self.request_timeout,
+        )
+        return batch, response
+
+    def _drop_sent_batch(self, sent_batch: Batch) -> None:
+        # we don't override createLock(), so lock is not None
+        with self.lock:  # type: ignore
+            # The previous lost count has been accounted by the server:
+            self.messages_buffer.dropped -= sent_batch.lost
+            # Number of messages dropped while we were busy flushing:
+            dropped_inadvertently = self.messages_buffer.head_serial_no - sent_batch.head_serial_no
+            remaining = len(sent_batch.logs) - dropped_inadvertently
+            if remaining > 0:
+                # Account for all the messages in the batch that were considered dropped:
+                self.messages_buffer.dropped -= dropped_inadvertently
+                # Drop the remainder:
+                self.messages_buffer.drop(remaining)
+            elif remaining < 0:
+                # Account for all the messages in the batch that were considered dropped:
+                self.messages_buffer.dropped -= len(sent_batch.logs)
+                # Uh oh! We lost some messages. The dropped count now should match -remaining.
+                # It will be reported to the server at the next flush.
+
+    def get_metadata(self) -> dict:
```

Comment:
I don't think that inheritence is the nicest way to pass this - why not just pass as parameter to ctor? What are the benefits of this approach?